src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -951,20 +947,23 @@ private void RemoveDocumentCore(DocumentId documentId, TextDocumentKind document
                 throw new ArgumentNullException(nameof(documentId));
             }
 
-            var document = this.CurrentSolution.GetDocument(documentId);
-            var hierarchy = this.GetHierarchy(documentId.ProjectId);
+            var document = this.CurrentSolution.GetTextDocument(documentId);
             if (document != null)
             {
+                var hierarchy = this.GetHierarchy(documentId.ProjectId);
+                Contract.ThrowIfNull(hierarchy, "Removing files from projects without hierarchies are not supported.");
```

Comment:
>Contract.ThrowIfNull(hierarchy, "Removing files from projects without hierarchies are not supported."); [](start = 16, length = 103)

Does the text add any value? Wouldn't a comment suffice?

Reply:
It means if you get a Watson you don't have to chase which ThrowIfNull failed. This could also get thrown (technically) through a call to TryApplyChanges, so if somebody with an extension files a bug it's also more actionable.

Reply:
Is this in any way different from hundreds of other places where we use Contract.ThrowIfNull w/o a message? How do you decide when to add the message?

In any case, if you think the message is useful I'd make it conditional, so that the string isn't loaded unless this actually fails 

```C#
if (hierarchy == null) { throw new InvalidOperationException("Removing files from projects without hierarchies are not supported."); }
```

Reply:
Is the cost of the load significant, especially given this doing COM interop and DTE? Are we doing an effort to find other ones that should be _removed_ across our codebase?

> Is this in any way different from hundreds of other places where we use Contract.ThrowIfNull w/o a message?

If nothing else, there's multiple ThrowIfNulls in the same method, so this would ease debugging if somebody gives me a stack that I have to debug. And honestly I'd value the debuggability of the code far more, especially since I can imagine this exception could leak if something is really busted.

Reply:
I understand why you want this for this particular case - here the overhead is indeed negligible. I'm just worried about potential widespread use of this pattern all over the code base without realizing it adds overhead. I'd rather remove the overload of `Contract.ThrowIfNull` that takes a message entirely. In places where the value of having the message is significant you can still write:

```C#
if (hierarchy == null) { throw new InvalidOperationException("Removing files from projects without hierarchies are not supported."); }
```

Reply:
Actually, I just realized that JIT might be smart and optimize it (not load the string until it's needed), if we make sure ThrowIfNull is inlined `MethodImpl(MethodImplOptions.AggressiveInlining)`. 
Tried it out [here](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxNMAfAAgJgIwCwAUFgAwAEWeAdAEoCuAdjAJZhzUDCA9mAA4sANogDKiAG4sAxnCgBuYsVyU8AdmIBvYuR3kA2gFk4MABbcAJgEl+ggBRHTF630EB5Pq26Mo1AIIBzfwRZKBZxOEtGQRZGGP8ASgBdbV0qADZKABZyThM4KQBrW25gACt8mHIS0qQVClgEeJSdLSJddvIWADNyYrLyAF4B8kZ6QUEmto6W5undUwRuAHcRuBXfBH96dmYAUQAPGQ8WL1sG+IUp6YBfWdur8lnDYzMrG3sXpxt3T29qADluJForFGAlkg90llyAAxWwxSr7SbtVrte6zKFYbIGCAxWxUUh6RLkCCbKDI3SouZwvAXWbtXL5Iqk/xQWoAIn2XK57LpD3acJwfLRxHuQA) and seems to work well.

So I take back my objection against using this pattern.

Reply:
@tmat Should I add that flag to our method then?

Reply:
It could use other improvements. Let me do that in a separate PR. I need to clean up some related code anyways.

Reply:
Resolving in favor of #42114 then.